### DISCUSSION TITLE
How to fly crazyflie in the wild?

### DISCUSSION THREAD

**Discussion by allegorywrite** [bitcraze/discussions#1032] [Q&A]
We are currently working on a project to enable crazyflie to fly autonomously outdoors.
As mentioned in the [previous article](https://forum.bitcraze.io/viewtopic.php?t=4266), the ToF sensor seems insufficient in the rough outdoors, so I am thinking of integrating VIO (Visual Inertial Odometry) as a solution.
However, the state estimator is embedded within the Cascaded PID controller's loop and I have no idea how to integrate it.
We are considering the following hardware options:
1. Single esp32-cam (currently the camera communicates with the laptop via wifi/bluetooth and has no integration with crazyflie other than power supply)
2. ai deck

Personally, I think it is best to process the video acquired by the AI deck on board and estimate state by VIO. However, due to the performance of GAP8 and compatibility with machine learning modules such as opencv, I am unsure whether AI deck is sufficient for VIO.
Furthermore, looking at the document, there is only an example of using video obtained from AI deck over WiFi, and it is unclear whether video information can be used for onboard internal control.
Considering this information, is it possible to integrate VIO into crazyflie?
If you have any tips for flying crazyflie outdoors, I would appreciate it if you could share them.
Thank you for your support in advance.

**Comment by knmcguire** [comment]
Hi! you are right, the tof sensor is sensitive to infrared light so it would make it highly unsuitable for outdoor flight.

Currently the only paper I can think off that did VOI (offboard) is an old workshop paper still with the Crazyflie 1: https://cvg.cit.tum.de/_media/spezial/bib/dunkley14iros.pdf

Any vision based papers on the aideck or other decks that do onboard processing are:

- Ultra-low Power Deep Learning-based Monocular Relative Localization Onboard Nano-quadrotors. Bonato et al 2023.
- Nanoflownet: Real-time dense optical flow on a nano quadcopter. Bouwmeester et al (2023)
- Fully Onboard AI-powered Human-Drone Pose Estimation on Ultra-low Power Autonomous Flying Nano-UAVs. Palossi et al (2021)

Not full VOI onboard yet but seeing these results on the limited GAP8 would make me think that at one point a form of visual based odometry should be possible! But so far no paper has been published yet on that.

Offboard ofcourse provides more possibilities and usage of existing libraries like OpenCV or the ROS based libraries. If you get a position estimate from those offboard, I recommend you to take a look at the [external position send in our python library](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/extpos/). We use this same structure for sending offboared MoCap data to the crazyflie so I would imagine it serving the same purpose for estimating position based on VIO.

Also the state estimate and the controllers are completely separate, so if you'd like to implement something onboard specifically to integrate odometry that is definitely possible with the extended kalman filter. So once you have figured out more of a plan for this, I recommend starting a discussion with just that question of how to integrate it. We prefer to work with one question per discussion if that is okay 😄

**Reply by allegorywrite** [reply]
Thanks for the informative and knowledgeable answer!
For me, processing video off board looks like a great solution. I'll set up another discussion about it later.

**Reply by knmcguire** [reply]
Awesome :) let us know if you need anymore help!
